curp/src/server/cmd_worker/conflict_checked_mpmc.rs

```diff
@@ -541,25 +540,59 @@ pub(in crate::server) fn channel<C: 'static + Command, CE: 'static + CommandExec
     let (filter_tx, recv_rx) = flume::unbounded();
     // recv from user to mark a msg done
     let (done_tx, done_rx) = flume::unbounded::<(Task<C>, bool)>();
-    let _ig = tokio::spawn(async move {
-        let mut filter = Filter::new(filter_tx, ce);
-        #[allow(clippy::integer_arithmetic, clippy::pattern_type_mismatch)]
-        // tokio internal triggers
-        loop {
-            tokio::select! {
-                biased; // cleanup filter first so that the buffer in filter can be kept as small as possible
-                Ok((task, succeeded)) = done_rx.recv_async() => {
-                    filter.progress(task.vid, succeeded);
-                },
-                Ok(event) = filter_rx.recv_async() => {
-                    filter.handle_event(event);
-                },
-                else => {
-                    error!("mpmc channel stopped unexpectedly");
-                    return;
+    let _ig = tokio::spawn(conflict_checked_mpmc_task(
+        filter_tx,
+        filter_rx,
+        ce,
+        shutdown_trigger,
+        done_rx,
+    ));
+    (CEEventTx(send_tx), recv_rx, done_tx)
+}
+
+/// Conflict checked mpmc task
+async fn conflict_checked_mpmc_task<C: 'static + Command, CE: 'static + CommandExecutor<C>>(
+    filter_tx: flume::Sender<Task<C>>,
+    filter_rx: flume::Receiver<CEEvent<C>>,
+    ce: Arc<CE>,
+    shutdown_trigger: shutdown::Trigger,
+    done_rx: flume::Receiver<(Task<C>, bool)>,
+) {
+    let mut shutdown_listener = shutdown_trigger.subscribe();
+    let mut filter = Filter::new(filter_tx, ce);
+    let mut is_shutdown_state = false;
+    #[allow(clippy::integer_arithmetic, clippy::pattern_type_mismatch)] // tokio internal triggers
+    loop {
+        tokio::select! {
+            biased; // cleanup filter first so that the buffer in filter can be kept as small as possible
+            sig = shutdown_listener.wait() => {
```

Comment:
can we add `; if !is_shutdown_state` here to prevent a poll of shutdown_listener?

Reply:
Fixed